File: modules/nf-core/soupx/templates/soupx.py
#!/usr/bin/env python3

import platform

import anndata2ri
import rpy2
import rpy2.robjects as ro
import scanpy

soupx = ro.packages.importr("SoupX")


def format_yaml_like(data: dict, indent: int = 0) -> str:
    """Formats a dictionary to a YAML-like string.

    Args:
        data (dict): The dictionary to format.
        indent (int): The current indentation level.

    Returns:
        str: A string formatted as YAML.
    """
    yaml_str = ""
    for key, value in data.items():
        spaces = "    " * indent
        if isinstance(value, dict):
            yaml_str += f"{spaces}{key}:\\n{format_yaml_like(value, indent + 1)}"
        else:
            yaml_str += f"{spaces}{key}: {value}\\n"
    return yaml_str


adata = scanpy.read_h5ad("${h5ad_filtered}")

adata_pp = adata.copy()
scanpy.pp.normalize_per_cell(adata_pp)
scanpy.pp.log1p(adata_pp)

scanpy.pp.pca(adata_pp)
scanpy.pp.neighbors(adata_pp)
scanpy.tl.leiden(adata_pp, key_added="soupx_groups")

soupx_groups = anndata2ri.py2rpy(adata_pp.obs["soupx_groups"])
del adata_pp

sce = anndata2ri.py2rpy(adata)

adata_raw = scanpy.read_h5ad("${h5ad_raw}")
sce_raw = anndata2ri.py2rpy(adata_raw)

get_counts = ro.r("function(sce) { assay(sce, 'X') }")

data = get_counts(sce)
data_raw = get_counts(sce_raw)

sc = soupx.SoupChannel(data_raw, data, calcSoupProfile=False)

soup_prof = ro.r(
    "function(data) { data.frame(row.names = rownames(data), est = rowSums(data)/sum(data), counts = rowSums(data)) }"
)(data)
sc = soupx.setSoupProfile(sc, soup_prof)
sc = soupx.setClusters(sc, soupx_groups)
sc = soupx.autoEstCont(sc, doPlot=False)
out = soupx.adjustCounts(sc, roundToInt=False)

adata.layers["ambient"] = anndata2ri.rpy2py(out).T

adata.write_h5ad("${prefix}.h5ad")

# versions
versions = {
    "${task.process}": {
        "python": platform.python_version(),
        "scanpy": scanpy.__version__,
        "anndata2ri": anndata2ri.__version__,
        "rpy2": rpy2.__version__,
        "SoupX": soupx.__version__,
    }
}

with open("versions.yml", "w") as f:
    f.write(format_yaml_like(versions))

File: modules/nf-core/soupx/templates/soupx.R
#!/usr/bin/env Rscript

suppressPackageStartupMessages({
    library(anndataR)
    library(Seurat)
    library(Matrix)
    library(SoupX)
})

set.seed(0)

# Template-interpolated by main.nf
h5ad_filtered     <- "${h5ad_filtered}"
h5ad_raw          <- "${h5ad_raw}"
prefix            <- "${prefix}"
npcs              <- ${npcs}
cluster_algorithm <- ${cluster_algorithm}

adata     <- read_h5ad(h5ad_filtered)
adata_raw <- read_h5ad(h5ad_raw)

# anndata stores cells x genes; Seurat/SoupX expect genes x cells
counts_filt <- t(adata\$X)
counts_raw  <- t(adata_raw\$X)

if (sum(counts_filt) == 0) stop("Filtered counts matrix is empty - cannot run SoupX.")

srt <- CreateSeuratObject(counts = counts_filt)
srt <- NormalizeData(srt, normalization.method = "LogNormalize", verbose = FALSE)
srt <- FindVariableFeatures(srt, verbose = FALSE)
srt <- ScaleData(srt, verbose = FALSE)
srt <- RunPCA(srt, npcs = npcs, verbose = FALSE)
srt <- FindNeighbors(srt, dims = 1:npcs, verbose = FALSE)
srt <- FindClusters(srt, algorithm = cluster_algorithm, verbose = FALSE)

soupx_groups <- setNames(as.character(Idents(srt)), colnames(srt))
rm(srt); gc()

sc <- SoupChannel(counts_raw, counts_filt, calcSoupProfile = FALSE)
soup_prof <- data.frame(
    row.names = rownames(counts_filt),
    est       = rowSums(counts_filt) / sum(counts_filt),
    counts    = rowSums(counts_filt)
)
sc  <- setSoupProfile(sc, soup_prof)
sc  <- setClusters(sc, soupx_groups)
sc  <- autoEstCont(sc, doPlot = FALSE)
out <- adjustCounts(sc, roundToInt = FALSE)

adata\$layers[["ambient"]] <- t(out)
write_h5ad(adata, paste0(prefix, ".h5ad"))

# versions.yml
versions <- c(
    '"${task.process}":',
    paste0("    r-base: ",     as.character(getRversion())),
    paste0("    soupx: ",      as.character(packageVersion("SoupX"))),
    paste0("    anndatar: ",   as.character(packageVersion("anndataR"))),
    paste0("    seurat: ",     as.character(packageVersion("Seurat"))),
    paste0("    leidenbase: ", as.character(packageVersion("leidenbase")))
)
writeLines(versions, "versions.yml")
